src/coreclr/src/jit/morph.cpp

```diff
@@ -7298,306 +7283,724 @@ GenTree* Compiler::fgMorphPotentialTailCall(GenTreeCall* call)
+    }
 
-    // First move the this pointer (if any) onto the regular arg list
-    GenTree* thisPtr = NULL;
+    // Put 'this' in normal param list
     if (call->gtCallThisArg != nullptr)
     {
+        JITDUMP("Moving this pointer into arg list\n");
+        GenTree* thisPtr    = nullptr;
         GenTree* objp       = call->gtCallThisArg->GetNode();
         call->gtCallThisArg = nullptr;
 
-        if ((call->gtFlags & GTF_CALL_NULLCHECK) || call->IsVirtualVtable())
+        if (call->NeedsNullCheck())
         {
-            thisPtr      = gtClone(objp, true);
+            // clone "this" if "this" has no side effects.
+            if ((thisPtr == nullptr) && !(objp->gtFlags & GTF_SIDE_EFFECT))
+            {
+                thisPtr = gtClone(objp, true);
+            }
+
             var_types vt = objp->TypeGet();
-            if (thisPtr == NULL)
+            if (thisPtr == nullptr)
             {
-                // Too complex, so use a temp
+                // create a temp if either "this" has side effects or "this" is too complex to clone.
+
+                // tmp = "this"
                 unsigned lclNum = lvaGrabTemp(true DEBUGARG("tail call thisptr"));
                 GenTree* asg    = gtNewTempAssign(lclNum, objp);
-                if (!call->IsVirtualVtable())
-                {
-                    // Add an indirection to get the nullcheck
-                    GenTree* tmp = gtNewLclvNode(lclNum, vt);
-                    GenTree* ind = gtNewOperNode(GT_IND, TYP_INT, tmp);
-                    asg          = gtNewOperNode(GT_COMMA, TYP_VOID, asg, ind);
-                }
-                objp    = gtNewOperNode(GT_COMMA, vt, asg, gtNewLclvNode(lclNum, vt));
-                thisPtr = gtNewLclvNode(lclNum, vt);
+
+                // COMMA(tmp = "this", deref(tmp))
+                GenTree* tmp = gtNewLclvNode(lclNum, vt);
+                GenTree* ind = gtNewOperNode(GT_IND, TYP_INT, tmp);
+                asg          = gtNewOperNode(GT_COMMA, TYP_VOID, asg, ind);
+
+                // COMMA(COMMA(tmp = "this", deref(tmp)), tmp)
+                thisPtr = gtNewOperNode(GT_COMMA, vt, asg, gtNewLclvNode(lclNum, vt));
             }
-            else if (!call->IsVirtualVtable())
+            else
             {
+                // thisPtr = COMMA(deref("this"), "this")
                 GenTree* ind = gtNewOperNode(GT_IND, TYP_INT, thisPtr);
-                objp         = gtNewOperNode(GT_COMMA, vt, ind, objp);
-                thisPtr      = gtClone(thisPtr, true);
+                thisPtr      = gtNewOperNode(GT_COMMA, vt, ind, gtClone(objp, true));
             }
 
             call->gtFlags &= ~GTF_CALL_NULLCHECK;
         }
+        else
+        {
+            thisPtr = objp;
+        }
 
-        call->gtCallArgs = gtPrependNewCallArg(objp, call->gtCallArgs);
+        // During rationalization tmp="this" and null check will
+        // materialize as embedded stmts in right execution order.
```

Comment:
The reference to embedded stmts (presumably from an existing comment) is obsolete. I would just say "will be materialized in the right execution order."